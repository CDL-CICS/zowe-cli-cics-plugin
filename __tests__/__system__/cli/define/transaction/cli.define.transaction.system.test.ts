/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import { TestEnvironment } from "../../../../__src__/environment/TestEnvironment";
import { ITestEnvironment } from "../../../../__src__/environment/doc/response/ITestEnvironment";
import { generateRandomAlphaNumericString, runCliScript } from "../../../../__src__/TestUtils";
import { Session } from "@brightside/imperative";
import { CicsCmciRestClient, CicsCmciConstants } from "../../../../../src";

let TEST_ENVIRONMENT: ITestEnvironment;
let regionName: string;
let csdGroup: string;


describe("CICS define transaction command", () => {

    beforeAll(async () => {
        TEST_ENVIRONMENT = await TestEnvironment.setUp({
            testName: "define_transaction",
            installPlugin: true,
            tempProfileTypes: ["cics"]
        });
        csdGroup = TEST_ENVIRONMENT.systemTestProperties.cmci.csdGroup;
        regionName = TEST_ENVIRONMENT.systemTestProperties.cmci.regionName;
    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENVIRONMENT);
    });

    const deleteTransaction = async (transactionName: string) => {
        const cmciProperties = TEST_ENVIRONMENT.systemTestProperties.cmci;
        const session = new Session({
            type: "basic",
            hostname: cmciProperties.host,
            port: cmciProperties.port,
            user: cmciProperties.user,
            password: cmciProperties.password,
            strictSSL: false,
            protocol: "http",
        });

        return CicsCmciRestClient.deleteExpectParsedXml(session,
            `/${CicsCmciConstants.CICS_SYSTEM_MANAGEMENT}/${CicsCmciConstants.CICS_DEFINITION_TRANSACTION}/${cmciProperties.regionName}` +
            `?CRITERIA=(NAME=${transactionName})&PARAMETER=CSDGROUP(${cmciProperties.csdGroup})`);
    };

    it("should be able to display the help", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_transaction_help.sh", TEST_ENVIRONMENT, []);
        expect(output.stderr.toString()).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toMatchSnapshot();
    });

    it("should be able to successfully define a transaction with basic options", async () => {
        const transactionNameSuffixLength = 3;
        const transactionName = "X" + generateRandomAlphaNumericString(transactionNameSuffixLength);
        const dummyPgmName = "TESTING";
        const output = runCliScript(__dirname + "/__scripts__/define_transaction.sh", TEST_ENVIRONMENT,
            [transactionName, dummyPgmName, csdGroup, regionName]);
        const stderr = output.stderr.toString();
        expect(stderr).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toContain("success");
        await deleteTransaction(transactionName);
    });

    it("should get a syntax error if transaction name is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_transaction.sh", TEST_ENVIRONMENT,
            ["", "FAKEPGM", "FAKEGRP", "FAKERGN"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("Missing Positional Option");
        expect(stderr).toContain("transactionName");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if program name is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_transaction.sh", TEST_ENVIRONMENT,
            ["FAKETRAN", "", "FAKEGRP", "FAKERGN"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("Missing Positional Option");
        expect(stderr).toContain("programName");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if CSD group is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/define_transaction.sh", TEST_ENVIRONMENT,
            ["FAKETRAN", "FAKEPGM", "", "FAKERGN"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("Missing Positional Option");
        expect(stderr).toContain("csdGroup");
        expect(output.status).toEqual(1);
    });

});
