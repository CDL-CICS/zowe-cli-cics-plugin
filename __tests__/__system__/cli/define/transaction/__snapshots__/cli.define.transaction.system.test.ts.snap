// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CICS define transaction command should be able to display the help 1`] = `
"
 COMMAND NAME
 ------------

   transaction | tran

 DESCRIPTION
 -----------

   Define a new transaction to CICS.

 USAGE
 -----

   bright cics define transaction <transactionName> <programName> <csdGroup> [options]

 POSITIONAL ARGUMENTS
 --------------------

   transactionName		 (string)

      The name of the new transaction to define. The maximum length of the transaction
      name is four characters.

   programName		 (string)

      The name of the program that the transaction uses. The maximum length of the
      program name is eight characters.

   csdGroup		 (string)

      The CICS system definition (CSD) Group for the new transaction that you want to
      define. The maximum length of the group name is eight characters.

 OPTIONS
 -------

   --region-name  (string)

      The CICS region name to which to define the new transaction

   --cics-plex  (string)

      The name of the CICSPlex to which to define the new transaction

 PROFILE OPTIONS
 ---------------

   --cics-profile  | --cics-p (string)

      The name of a (cics) profile to load for this command execution.

 GLOBAL OPTIONS
 --------------

   --response-format-json  | --rfj (boolean)

      Produce the command response as a JSON document

   --help  | -h (boolean)

      Display help text


 EXAMPLES
 --------

   - Define a transaction named TRN1 for the program named
   PGM123 to the region named MYREGION in the CSD group MYGRP:

      $ bright cics define transaction TRN1 PGM123 MYGRP --region-name MYREGION

"
`;
